### PR TITLE
Fix Command line interface commands and connection management service to prompt for provider installation

### LINKED ISSUE #25050: openConnectionDialog URI Handler doesn't prompt to install missing extension
1. Open ADS via URI using the `openConnectionDialog` path and a provider specified that isn't installed - e.g. `azuredatastudio://openConnectionDialog?server=myserver.mysql.database.azure.com&provider=MySQL&user=sa`

Expected : User is prompted to install the missing extension

Actual : The user is only prompted if they want to open the connection, but then the connection dialog just defaults to the MSSQL provider since the one specified doesn't exist.

### LINKED ISSUE #25055: Support ‘Connect’ command to work from Command Line interface
- Azure Data Studio Version: main
- OS Version: Windows 11

Execute command from command prompt:
`azuredatastudio --command connect --user sa --server localhost`

Spent some time while working on https://github.com/microsoft/azuredatastudio/issues/25050, I noticed that prompt to 'connect' doesn't work when invoked from command line, and threw this error instead. It doesn't seem to cause issues when invoked from URL:

```
[error] command 'connect' not found: Error: command 'connect' not found
    at S._tryExecuteCommand (vscode-file://vscode-app/c:/Users/cmalhotra/AppData/Local/Programs/Azure%20Data%20Studio/resources/app/out/vs/workbench/workbench.desktop.main.js:2322:17512) [<root>]
    at S.executeCommand (vscode-file://vscode-app/c:/Users/cmalhotra/AppData/Local/Programs/Azure%20Data%20Studio/resources/app/out/vs/workbench/workbench.desktop.main.js:2322:17394) [<root>]
    at async C.processCommandLine (vscode-file://vscode-app/c:/Users/cmalhotra/AppData/Local/Programs/Azure%20Data%20Studio/resources/app/out/vs/workbench/workbench.desktop.main.js:2097:19922) [<root>]
```

It seems as though this command got removed at some point, whereas the code to call the command stayed..
As the code that invoked this command was added quite long ago in PR https://github.com/microsoft/azuredatastudio/issues/4364 and hadn't been modified in a long time..

cc @Charles-Gagnon @alanrenmsft for awareness.

### PR DESCRIPTION
**Pull request by cheenamalhotra**
Fixes #25050 and fixes #25055

Tested both commands from CLI args when working with unknown/known providers:

connect-default
openConnection-default
Note: Prompt added in Connection Management Service asks users to install extension again when user tries to connect despite canceling the first installation prompt.
cli-fixed

### CODE REVIEW

File: src/sql/workbench/contrib/commandLine/electron-sandbox/commandLine.ts [hunk 1]
```diff
@@ -118,37 +118,27 @@ export class CommandLineWorkbenchContribution implements IWorkbenchContribution,
...
+					await this._connectionManagementService.connectIfNotConnected(profile, args.showDashboard ? 'dashboard' : 'connection', true);
+					// Before sending to extensions, we should a) serialize to IConnectionProfile or things will fail,
+					// and b) use the latest version of the profile from the service so most fields are filled in.
+					let updatedProfile = this._connectionManagementService.getConnectionProfileById(profile.id);
+					connectedContext = { connectionProfile: new ConnectionProfile(this._capabilitiesService, updatedProfile).toIConnectionProfile() };
+				} catch (err) {
+					this.logService.warn('Failed to connect due to error: ' + getErrorMessage(err));
 				}
-				// Recreate the profile here now that we have our provider registered so that we get the correct
-				// option names. See https://github.com/microsoft/azuredatastudio/issues/20773 for details about the issue
```
Comment (cheenamalhotra) [inline src/sql/workbench/contrib/commandLine/electron-sandbox/commandLine.ts:130]: @Charles-Gagnon 
To resolve this, I'm letting the provider be handled first, before profile is created. Let me know if I'm missing something..

File: src/sql/workbench/contrib/commandLine/electron-sandbox/commandLine.ts [hunk 2]
```diff
@@ -118,37 +118,27 @@ export class CommandLineWorkbenchContribution implements IWorkbenchContribution,
...
-				this.logService.warn('Failed to connect due to error: ' + getErrorMessage(err));
 			}
-		}
-		if (commandName) {
 			if (this._notificationService) {
 				this._notificationService.status(localize('runningCommandLabel', "Running command: {0}", commandName), { hideAfter: 2500 });
 			}
-			await this._commandService.executeCommand(commandName, connectedContext);
-		} else if (connectedContext) {
+			await this.runCommandHandler(commandName, args);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/commandLine/electron-sandbox/commandLine.ts:139]: This is meant to allow users to run a command immediately upon startup so we ideally shouldn't be removing that functionality. And I'm not sure about overloading "command" to be both "hardcoded" commands in this file and the registered commands...so maybe we can find a different name for the hardcoded ones (action perhaps?)

I forget exactly why we added that in the first place - it was added in https://github.com/microsoft/azuredatastudio/pull/3690 @shueybubbles do you recall what the exact use case for this was?

Comment (shueybubbles) [reply src/sql/workbench/contrib/commandLine/electron-sandbox/commandLine.ts:139]: this is how SSMS opens ADS to the New Notebook experience

From the object explorer XML:
```xml
    <Object name='NewNotebook' base='StartAzureDataStudio'>
      <Properties>
        <LocalizableProperty Name='MenuText'>New &amp;Notebook</LocalizableProperty>
        <Property Name="AdditionalArguments">--command=notebook.command.new</Property>
```

File: src/sql/workbench/contrib/commandLine/electron-sandbox/commandLine.ts [hunk 3]
```diff
@@ -118,37 +118,27 @@ export class CommandLineWorkbenchContribution implements IWorkbenchContribution,
...
 		}
-		let connectedContext: azdata.ConnectedContext = undefined;
-		if (profile) {
-			if (profile.providerName && !this._capabilitiesService.providers[profile.providerName]) {
-				const installed = await this._connectionManagementService.handleUnsupportedProvider(profile.providerName);
-				if (!installed) {
-					// User cancelled install prompt so exit early since we won't be able to connect
-					return;
+		if (commandName) {
+			if (profile && commandName === Command.connect) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/commandLine/electron-sandbox/commandLine.ts:122]: Maybe this logic should be reversed, with the connect handler only handling the connect-specific logic, and have the handleURL function go through processCommandLine directly instead (which then calls the appropriate command handler). It just seems like it'd be cleaner to have the command-specific logic each in their own handler.

File: src/sql/workbench/services/connection/browser/connectionDialogService.ts [hunk 4]
```diff
@@ -383,16 +383,24 @@ export class ConnectionDialogService implements IConnectionDialogService {
...
 	}
 
-	private async createModel(model: IConnectionProfile): Promise<ConnectionProfile> {
+	private async createModel(model: IConnectionProfile): Promise<ConnectionProfile | undefined> {
 		const defaultProvider = this.getDefaultProviderName();
+		// Handle unsupported provider
+		if (model?.providerName && !this._capabilitiesService.providers[model.providerName]) {
+			const installed = await this._connectionManagementService.handleUnsupportedProvider(model.providerName);
+			if (!installed) {
+				// User cancelled install prompt so exit early since we won't be able to connect
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/connection/browser/connectionDialogService.ts:400]: What is the behavior with this - does the dialog just not open? Or does it open but is blank? I think my slight preference is the first - although if we do open it I would at least suggest putting in what info we can (and just let the provider default to MSSQL)

File: src/sql/workbench/contrib/commandLine/electron-sandbox/commandLine.ts (hunk 1 above)
Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/commandLine/electron-sandbox/commandLine.ts:130]: Seems like a good idea to me! And yeah, that should fix this issue I believe

File: src/sql/workbench/contrib/commandLine/electron-sandbox/commandLine.ts (hunk 2 above)
Comment (cheenamalhotra) [reply src/sql/workbench/contrib/commandLine/electron-sandbox/commandLine.ts:139]: Makes sense, thanks for context. Will update PR accordingly.

File: src/sql/workbench/services/connection/browser/connectionDialogService.ts (hunk 4 above)
Comment (cheenamalhotra) [reply src/sql/workbench/services/connection/browser/connectionDialogService.ts:400]: Dialog doesn't open if user cancels the prompt.

Comment (Charles-Gagnon) [reply src/sql/workbench/services/connection/browser/connectionDialogService.ts:400]: Sounds good, that seems like reasonable behavior for now.

File: src/sql/workbench/contrib/commandLine/electron-sandbox/commandLine.ts [hunk 5]
```diff
@@ -171,53 +171,71 @@ export class CommandLineWorkbenchContribution implements IWorkbenchContribution,
...
-		path: 'connect'
+		path: Command.connect
 	})
-	public async handleConnect(uri: URI): Promise<boolean> {
+	public async handleConnect(args: NativeParsedArgs): Promise<boolean> {
 		try {
-			let args = this.parseProtocolArgs(uri);
 			if (!args.server) {
-				this._notificationService.warn(localize('warnServerRequired', "Cannot connect as no server information was provided"));
+				this._notificationService?.warn(localize('warnServerRequired', "Cannot connect as no server information was provided"));
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/commandLine/electron-sandbox/commandLine.ts:208]: We should never have an undefined notification service...if this is for tests then I would suggest using this : https://github.com/Microsoft/azuredatastudio/blob/main/src/vs/platform/notification/test/common/testNotificationService.ts#L10

File: src/sql/workbench/contrib/commandLine/electron-sandbox/commandLine.ts [hunk 6]
```diff
@@ -171,53 +171,71 @@ export class CommandLineWorkbenchContribution implements IWorkbenchContribution,
...
+						showConnectionDialogOnError: true,
+						showFirewallRuleOnError: true
+					});
+				} catch (err) {
+					this.logService.warn('Failed to connect due to error: ' + getErrorMessage(err));
+				}
 			}
 		} catch (err) {
-			this._notificationService.error(localize('errConnectUrl', "Could not open URL due to error {0}", getErrorMessage(err)));
+			this._notificationService?.error(localize('errConnectUrl', "Could not open URL due to error {0}", getErrorMessage(err)));
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/commandLine/electron-sandbox/commandLine.ts:226]: Same here, this should never be undefined

File: src/sql/workbench/contrib/commandLine/electron-sandbox/commandLine.ts (hunk 5 above)
Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/commandLine/electron-sandbox/commandLine.ts:208]: (same for all the other places you do that in this file)

File: src/sql/workbench/contrib/commandLine/test/electron-browser/commandLine.test.ts [hunk 7]
```diff
@@ -178,6 +182,27 @@ suite('commandLineService tests', () => {
...
+			.returns(() => TypeMoq.It.isAny())
+			.verifiable(TypeMoq.Times.once());
+		connectionManagementService.setup(c => c.getConnectionGroups(TypeMoq.It.isAny())).returns(() => TypeMoq.It.isAny());
+		connectionManagementService.setup(c => c.hasRegisteredServers()).returns(() => false);
+		connectionManagementService.setup(c => c.connectIfNotConnected(TypeMoq.It.isAny(), TypeMoq.It.isAny()))
+			.returns(() => new Promise<string>((resolve, reject) => { resolve('unused'); }))
+			.verifiable(TypeMoq.Times.never());
+		const configurationService = getConfigurationServiceMock(true);
+		let contribution = getCommandLineContribution(connectionManagementService.object, configurationService.object);
+		return contribution.processCommandLine(args).then(() => {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/commandLine/test/electron-browser/commandLine.test.ts:201]: Making this async would be much cleaner

File: src/sql/workbench/contrib/commandLine/electron-sandbox/commandLine.ts [hunk 8]
```diff
@@ -143,11 +136,18 @@ export class CommandLineWorkbenchContribution implements IWorkbenchContribution,
...
 			if (this._notificationService) {
 				this._notificationService.status(localize('runningCommandLabel', "Running command: {0}", commandName), { hideAfter: 2500 });
 			}
-			await this._commandService.executeCommand(commandName, connectedContext);
+			if (commandName === Command.connect || commandName === Command.openConnectionDialog) {
+				// Run handlers for 'connect' and 'openConnectionDialog' commands.
+				await this.runCommandHandler(commandName, args);
+			} else {
+				// Execute other commands via commandService.
+				await this._commandService.executeCommand(commandName, connectedContext);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/commandLine/electron-sandbox/commandLine.ts:149]: Make sure to update https://github.com/Microsoft/azuredatastudio/blob/main/src/vs/platform/environment/common/argv.ts#L145, https://github.com/Microsoft/azuredatastudio/blob/main/src/vs/platform/environment/node/argv.ts#L185 and the [docs](https://learn.microsoft.com/en-us/azure-data-studio/command-line#supported-commands) as well

(and if you haven't already it'd be good to go through and make sure the other help options are correct for the various fields)

**Review by Charles-Gagnon — APPROVED** [review]
Looks good, aside from making sure the help stuff/comments are updated.

I'd also suggest making sure to validate that the SSMS New Notebook action still works as expected after the updates.

**Comment by cheenamalhotra** [comment]
> Looks good, aside from making sure the help stuff/comments are updated.
> 
> I'd also suggest making sure to validate that the SSMS New Notebook action still works as expected after the updates.

New Notebook action is fine, but other commands are currently not able to consume connection profile passed by user, due to #25068 - let me know your thoughts on it. I was thinking to deprecate this context and use the same BaseActionContext to keep things simplified.

**Pull request merged**